test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java

```diff
@@ -0,0 +1,242 @@
+        }
+    }
+    public static class CustomSocketFactory extends SocketFactory {
+        private static CustomSocket customSocket;
+
+        public static CustomSocketFactory getDefault() {
+            return new CustomSocketFactory();
+        }
+
+        @Override
+        public Socket createSocket() throws SocketException {
+            customSocket = new CustomSocket();
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String s, int timeout) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String host, int port, InetAddress localHost,
+                                   int localPort) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress host, int port) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress address, int port,
+                                   InetAddress localAddress, int localPort) {
+            return customSocket;
+        }
+    }
+
+    private static class CustomSocket extends Socket {
+        private int closeMethodCalled = 0;
+        public CustomSocket () {
+            closeMethodCalled = 0;
+        }
+        public int closeMethodCalledCount() {
+            return closeMethodCalled;
+        }
+
+        @Override
+        public void close() throws java.io.IOException {
+            closeMethodCalled++;
+            super.close();
+        }
+    }
+
+    private static void setKeyStore() throws Exception {
+
+        String fileName = "ksWithSAN", dir = System.getProperty("test.src", ".") + File.separator;
+
+        System.setProperty("javax.net.ssl.keyStore", dir + fileName);
+        System.setProperty("javax.net.ssl.keyStorePassword", "welcome1");
+        System.setProperty("javax.net.ssl.trustStore", dir + fileName);
+        System.setProperty("javax.net.ssl.trustStorePassword", "welcome1");
+    }
+
+    static class TestServer extends Thread implements AutoCloseable {
+        private boolean isForceToSleep = false;
+        private final ServerSocket serverSocket;
+        private final int PORT;
+
+        TestServer(boolean isForceToSleep) {
+            this.isForceToSleep = isForceToSleep;
+            try {
+                SSLServerSocketFactory socketFactory = (SSLServerSocketFactory) SSLServerSocketFactory.getDefault();
+                serverSocket = socketFactory.createServerSocket(0, 0, InetAddress.getLoopbackAddress());
+                PORT = serverSocket.getLocalPort();
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+            setDaemon(true);
+        }
+
+        public int getPortNumber() {
+            return PORT;
+        }
+
+        @Override
+        public void run() {
+            try (Socket socket = serverSocket.accept(); InputStream in = socket.getInputStream();
```

Comment:
formatting: this can be made shorter by putting `InputStream in = socket.getInputStream();`  on a separate line